go/vt/discovery/topology_watcher.go

```diff
@@ -110,23 +110,52 @@ func (tw *TopologyWatcher) getTablets() ([]*topo.TabletInfo, error) {
 	return tw.topoServer.GetTabletsByCell(tw.ctx, tw.cell, nil)
 }
 
+func (tw *TopologyWatcher) getTabletsByShard(keyspace string, shard string) ([]*topo.TabletInfo, error) {
+	return tw.topoServer.GetTabletsByShardCell(tw.ctx, keyspace, shard, []string{tw.cell})
+}
+
 // Start starts the topology watcher.
 func (tw *TopologyWatcher) Start() {
-	tw.wg.Add(1)
+	tw.wg.Add(2)
+	// Goroutine to refresh the tablets list periodically.
 	go func(t *TopologyWatcher) {
 		defer t.wg.Done()
 		ticker := time.NewTicker(t.refreshInterval)
 		defer ticker.Stop()
 		for {
-			t.loadTablets()
+			// Since we are going to load all the tablets,
+			// we can clear out the entire list for reloading
+			// specific keyspace shards.
+			func() {
+				for {
+					select {
+					case <-tw.healthcheck.GetLoadTabletsTrigger():
+					default:
+						return
+					}
+				}
+			}()
+			t.loadTablets("", "")
 			select {
 			case <-t.ctx.Done():
 				return
-			case <-tw.healthcheck.GetLoadTabletsTrigger():
 			case <-ticker.C:
 			}
 		}
 	}(tw)
+	// Go routine to refresh tablets for a specific
+	// keyspace shard.
+	go func(t *TopologyWatcher) {
```

Comment:
Why does this need to be a separate goroutine? Why not
```
			case kss := <-t.healthcheck.GetLoadTabletsTrigger():
			    t.loadTablets(keyspace, shard)
			case <-ticker.C:
			    t.loadTablets()
```
We'll also have to add a call to `loadTablets` before starting the goroutine. The current version of the code was trying to be cute and avoid that, but it's not a big deal. Better than 2 goroutines and clearing out the channel etc.

Reply:
Fixed